docs/proposals/post-deployment-hooks.md

```diff
@@ -0,0 +1,223 @@
+
+Database migrations are complex and introduce downtime concerns. Here are [some examples](https://blog.rainforestqa.com/2014-06-27-zero-downtime-database-migrations) of zero-downtime Rails migration workflows.
+
+Deployments including database migrations must make special considerations:
+
+1.  Deployed code must be consistent with the currently deployed schema at any point in time.
+2.  Database or table locking must be minimized or eliminated to prevent service outages.
+
+The workflows which are effective at ensuring zero downtime migrations are typically multi-phased. For a user orchestrating a zero downtime migration deployment, it's likely the user needs to verify each deployment step discretely, with the option to abort and rollback after each phase.
+
+###### Container lifecycle hooks
+
+Container lifecycle hooks introduce problems with Rails migrations:
+
+1. The migration hook will execute once per pod in the deployment. This is a race and can result in concurrent migration attempts against a shared database. This burdens the hook author with coordinating the hooks or otherwise safeguarding the migration code to ensure it is safe for concurrent execution.
+  1. The uncoordinated aspect of these hooks is at also at odds with the general requirement to provide discrete step verification for the deployment.
+2. The migration hook will execute in the same pod as application containers, consuming resources allocated for the application.
+  1. This can cause instability, as it's unlikely the application pod resource allocation takes into account the temporarily increased requirements of a transient deployment step.
+
+###### Deployment hooks
+
+Deployment hooks satisfy this use case by providing a means to execute the hook only once per logical deployment. The hook is expressed as a run-once pod which provides the migration with its own resource allocation decoupled from the application.
+
+
+#### Use-case: Call a webhook
+
+Consider a Java application whose deployment should result in a webhook being called to notify some external system of the new deployment's availability.
+
+###### Container lifecycle hooks
+
+Container lifecycle hooks are inappropriate for this use case. The webhook should be called only once. Without a hook which coordinates with other replicated pod hook executions, the hook would:
+
+1. Be fired once per pod in the deployment.
+2. Be fired during the scale-up of the deployment rather than following the logical deployment as a whole.
+
+###### Deployment hooks
+
+A deployment hook configured to run following the scale-up of the deployment would satisfy the use case.
+
+
+## Proposed design
+
+Deployment hooks are implemented as run-once pods which can be executed at one or both of the following points during the deployment lifecycle:
```

Comment:
(this reply is meant to address not only the adjacent comment but also [this one](https://github.com/openshift/origin/pull/1224/files#r26335200) and [this one](https://github.com/openshift/origin/pull/1224/files#r26335243))

My thinking was that if your hook code was totally decoupled from your deployment image, you go down a path where you end up needing to capture all the pod template fields to run the hook anyway (first it's the command, then maybe you also want to inject special environment into your hook, you also might want to set the restart policy if it's retryable, etc.). And we already have an API for that input via PodTemplate. I agree it might be overkill, but we'll have to draw some line about what subset of fields we'd accept to ultimately jam into a PodTemplate.

What I have described currently could be 0 in your list:

0. Run run once pod during deployment process using **arbitrary** image and an executable in the image

If we're willing to narrow the capabilities to your 1-3 list, we can come up with a more special purpose and concise API.

So, do we scrap 0 and move on to figuring out how to provide the API for 1 and 2?